### ISSUE TITLE
Engine_ctrl commands support in provider

### ISSUE THREAD

**Issue by Yogaraj-Alamenda** [openssl/openssl#17182]
It seems OpenSSL 3.0 provider doesn’t have Engine Ctrl commands support programmatically from application which was there in OpenSSL 1.1.1 or before.

We are working in adding provider support to [qatengine](https://github.com/intel/QAT_Engine) which relies on Engine ctrl command for supporting features such External polling, heuristic polling, etc where application interacts with engine using the Engine ctrl API for polling and act as an interface between engine and application [(async_mode_nginx)](https://github.com/intel/asynch_mode_nginx) for set and get params to and from engine. 

Is there any other way that this could be achieved via provider as without engine ctrl support we will not be able to support  these features that engine framework supports.

Details about Features that use engine ctrl in [here  ](https://github.com/intel/asynch_mode_nginx#support-for-nginx-side-polling)

**Comment by mattcaswell** [comment]
Yes and no. There is a way to pass configuration parameters from a config file to a provider. There isn't a way to pass parameters dynamically for a running provider.

See the following man page for details (Section "Core parameters"):

https://www.openssl.org/docs/man3.0/man7/provider-base.html

**Comment by Yogaraj-Alamenda** [comment]
@mattcaswell Is there any plan to support it as we would need this support for the features mentioned above.

**Comment by paulidale** [comment]
Looks like we should have had a provider set params call to let this happen.
There is a very kludgy was to do this using get params which I'd not recommend.

**Comment by t8m** [comment]
IMO we can add that in 3.1

**Comment by t8m** [comment]
Now the question is if those set_params should be global per-provider or libctx specific. IMO they should be libctx specific.

**Comment by paulidale** [comment]
That would imply that the get params should also be libctx specific.  They are currently global but are limited to constant information.  Once we've got mutable state in providers, the get needs to be able to retrieve it.

Agreed, 3.1 is possible.

**Comment by paulidale** [comment]
Perhaps the libctx vs global state could be a decision for each provider?

**Comment by t8m** [comment]
> Perhaps the libctx vs global state could be a decision for each provider?

Perhaps. What I mean is that the provider should be able to recognize if the set_params() was associated with a particular libctx where it is loaded.

> Such parameters have nothing to do with the core.

I am not saying they have. I'm just saying that the provider needs to associate the call with a core handle.

**Comment by Yogaraj-Alamenda** [comment]
Just checking if there are any plans to implement this or any alternacive way in providers ?

**Comment by petrovr** [comment]
So long time without progress :(. 
Now I'm thinking how to use store URIs to set/get parameters at run-time.

**Comment by beldmit** [comment]
`OSSL_load_provider_ex` supports initialization of providers with an array of parameters. As it is intended to be compatible with config activation, it implies UTF-8 strings as input

**Comment by hardikpatel9** [comment]
To use provider, what is the equivalent of ENGINE_ctrl_cmd for set and retrieving custom data? How can we set and get such data within a provider from another application like Nginx, and is it possible to modify or access this data at runtime?

**Comment by t8m** [comment]
It is the OSSL_load_provider_ex() function as mentioned above. You can specify any parameters of a provider when loading the provider from the application.

**Comment by hardikpatel9** [comment]
"We are unable to locate the API OSSL_load_provider_ex() in the OpenSSL master branch or in version 3.5.0. Could you please provide further guidance or clarification regarding this API and example ?"

**Comment by t8m** [comment]
I am sorry for confusion. It is `OSSL_PROVIDER_load_ex()` actually - https://docs.openssl.org/master/man3/OSSL_PROVIDER/#name

**Comment by hardikpatel9** [comment]
Thanks @t8m for providing the API. Can we also use this API at runtime to load/update the parameter?  Also, to fetch/get this same, which API needs to be used?

**Comment by mattcaswell** [comment]
> Can we also use this API at runtime to load/update the parameter?

This API enables you to pass parameters on load of the provider. It is not currently possible to update a parameter for an already running provider.

> Also, to fetch/get this same, which API needs to be used?

Not really sure what you're asking here. You don't need to fetch anything. Just call `OSSL_PROVIDER_load_ex()` to load your provider.

**Comment by hardikpatel9** [comment]
Thanks @mattcaswell  for the reply,
 
We are using the Nginx application (https://github.com/intel/asynch_mode_nginx) and passing a custom parameter (e.g., ABC=1) from Nginx to the OpenSSL library (qatengine). Once the provider sets up the data, we need to retrieve those parameters later at runtime. Therefore, we are looking for a get or fetch API to retrieve the data previously set by OSSL_PROVIDER_load_ex().

Is there any plan in the future to implement an API for already loaded providers to update and pass the custom data?
